File: src/test/java/org/shsts/tinactory/datagen/builder/TechBuilder.java
package org.shsts.tinactory.datagen.builder;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import javax.annotation.Nullable;
import javax.annotation.ParametersAreNonnullByDefault;
import net.minecraft.MethodsReturnNonnullByDefault;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.ItemLike;
import net.minecraftforge.common.data.ExistingFileHelper;
import org.shsts.tinactory.api.tech.ITechnology;
import org.shsts.tinactory.core.builder.Builder;
import org.shsts.tinactory.core.electric.Voltage;
import org.shsts.tinactory.core.recipe.ProcessingIngredients;
import org.shsts.tinactory.datagen.provider.TechProvider;
import org.shsts.tinycorelib.api.core.ILoc;
import org.shsts.tinycorelib.datagen.api.IDataHandler;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Supplier;

import static org.shsts.tinactory.content.AllItems.getComponent;
import static org.shsts.tinactory.content.AllRecipes.RESEARCH_BENCH;
import static org.shsts.tinactory.test.TinactoryTest.DATA_GEN;

@ParametersAreNonnullByDefault
@MethodsReturnNonnullByDefault
public class TechBuilder<P> extends Builder<JsonObject, P, TechBuilder<P>> implements ILoc {
    public static final int RANK_PER_VOLTAGE = 1000;

    private final ResourceLocation loc;
    private final List<ResourceLocation> depends = new ArrayList<>();
    private long maxProgress = 0;
    private int rank = 0;
    private final Map<String, Integer> modifiers = new HashMap<>();
    @Nullable
    private Supplier<ResourceLocation> displayItem = null;
    @Nullable
    private ResourceLocation displayTexture = null;
    @Nullable
    private Voltage researchVoltage;

    public TechBuilder(P parent, ResourceLocation loc) {
        super(parent);
        this.loc = loc;
    }

    public static <P> TechBuilder<P> factory(IDataHandler<TechProvider> handler,
        P parent, ResourceLocation loc) {
        var builder = new TechBuilder<>(parent, loc);
        return builder.onBuild(() -> builder.onRegister(handler));
    }

    public TechBuilder<P> depends(ResourceLocation... loc) {
        depends.addAll(List.of(loc));
        return this;
    }

    public TechBuilder<P> maxProgress(long maxProgress) {
        this.maxProgress = maxProgress;
        return this;
    }

    public TechBuilder<P> displayItem(ResourceLocation loc) {
        displayItem = () -> loc;
        return this;
    }

    public TechBuilder<P> displayItem(ItemLike item) {
        return displayItem(() -> item);
    }

    public TechBuilder<P> displayItem(Supplier<? extends ItemLike> item) {
        displayItem = () -> item.get().asItem().getRegistryName();
        return this;
    }

    public TechBuilder<P> displayTexture(ResourceLocation val) {
        displayTexture = val;
        return this;
    }

    public TechBuilder<P> researchVoltage(Voltage val) {
        researchVoltage = val;
        return this;
    }

    public TechBuilder<P> modifier(String key, int val) {
        modifiers.put(key, val);
        return this;
    }

    public TechBuilder<P> rank(int val) {
        rank = val;
        return this;
    }

    @Override
    protected JsonObject createObject() {
        assert maxProgress > 0;
        var jo = new JsonObject();
        jo.addProperty("max_progress", maxProgress);
        var rank1 = researchVoltage != null ? RANK_PER_VOLTAGE * researchVoltage.rank : 0;
        jo.addProperty("rank", rank + rank1);
        var ja = new JsonArray();
        depends.forEach(d -> ja.add(d.toString()));
        jo.add("depends", ja);

        if (displayItem != null) {
            jo.addProperty("display_item", displayItem.get().toString());
        } else if (displayTexture != null) {
            jo.addProperty("display_texture", displayTexture.toString());
        }

        var jo1 = new JsonObject();
        for (var entry : modifiers.entrySet()) {
            jo1.addProperty(entry.getKey(), entry.getValue());
        }
        jo.add("modifiers", jo1);
        return jo;
    }

    public void validate(ExistingFileHelper existingFileHelper) {
        for (var loc : depends) {
            if (!existingFileHelper.exists(loc, TechProvider.RESOURCE_TYPE)) {
                throw new IllegalStateException("Technology at %s does not exist".formatted(loc));
            }
        }
    }

    @Override
    public ResourceLocation loc() {
        return loc;
    }

    private void onRegister(IDataHandler<TechProvider> handler) {
        var dataGen = handler.dataGen();
        handler.addCallback(p -> p.addTech(this));
        var description = ITechnology.getDescriptionId(loc);
        var details = ITechnology.getDetailsId(loc);
        dataGen.trackLang(description);
        dataGen.trackLang(details);

        if (researchVoltage != null) {
            var input = getComponent("research_equipment").get(researchVoltage).get();
            RESEARCH_BENCH.recipe(DATA_GEN, loc)
                .target(loc)
                .input(new ProcessingIngredients.ItemIngredient(new ItemStack(input, 1)))
                .voltage(researchVoltage.value)
                .power((long) (0.25 * researchVoltage.value))
                .workTicks(200)
                .build();
        }
    }

    public ResourceLocation register() {
        build();
        return loc;
    }
}

File: datagen/src/main/kotlin/org/shsts/tinactory/datagen/builder/TechBuilder.kt
package org.shsts.tinactory.datagen.builder

import net.minecraft.resources.ResourceLocation
import net.minecraft.world.level.ItemLike
import net.neoforged.neoforge.common.data.ExistingFileHelper
import org.shsts.tinactory.AllItems.getComponent
import org.shsts.tinactory.api.tech.ITechnology
import org.shsts.tinactory.core.builder.Builder
import org.shsts.tinactory.core.electric.Voltage
import org.shsts.tinactory.core.tech.Technology
import org.shsts.tinactory.core.util.LocHelper.modLoc
import org.shsts.tinactory.datagen.content.RegistryHelper
import org.shsts.tinactory.datagen.content.builder.RecipeFactories.research
import org.shsts.tinactory.datagen.content.component.item
import org.shsts.tinactory.datagen.provider.TechProvider
import org.shsts.tinycorelib.api.registrate.entry.IEntry
import org.shsts.tinycorelib.datagen.api.IDataHandler
import java.util.Optional

class TechBuilder<P>(parent: P) : Builder<Technology, P, TechBuilder<P>>(parent) {
    private val depends = mutableListOf<ResourceLocation>()
    private val modifiers = mutableMapOf<String, Int>()
    private var maxProgress = 0L
    private var rank = 0
    private var displayItem: ResourceLocation? = null
    private var displayTexture: ResourceLocation? = null
    private var voltage: Voltage? = null
    private var noResearch = false

    fun depends(vararg loc: ResourceLocation) {
        depends += loc
    }

    fun maxProgress(value: Long) {
        maxProgress = value
    }

    fun displayItem(loc: ResourceLocation) {
        displayItem = loc
    }

    fun displayItem(id: String) {
        displayItem = modLoc(id)
    }

    fun displayItem(item: ItemLike) {
        displayItem = RegistryHelper.itemLoc(item)
    }

    fun displayItem(item: IEntry<out ItemLike>) {
        displayItem = item.loc()
    }

    fun displayTexture(value: ResourceLocation) {
        displayTexture = value
    }

    fun researchVoltage(value: Voltage) {
        voltage = value
    }

    fun noResearch() {
        noResearch = true
    }

    fun modifier(key: String, value: Int) {
        modifiers[key] = value
    }

    fun rank(value: Int) {
        rank = value
    }

    override fun createObject(): Technology {
        return Technology(depends.toList(), maxProgress, modifiers.toMap(),
            Optional.ofNullable(displayItem),
            Optional.ofNullable(displayTexture),
            rank + RANK_PER_VOLTAGE * voltage!!.rank)
    }

    fun validate(existingFileHelper: ExistingFileHelper) {
        check(maxProgress > 0)
        checkNotNull(voltage)
        for (loc in depends) {
            check(existingFileHelper.exists(loc, TechProvider.RESOURCE_TYPE)) {
                "Technology at $loc does not exist"
            }
        }
    }

    private fun onRegister(handler: IDataHandler<TechProvider>, loc: ResourceLocation) {
        val dataGen = handler.dataGen()
        handler.addCallback { it.addTech(loc, this) }
        val description = ITechnology.getDescriptionId(loc)
        val details = ITechnology.getDetailsId(loc)
        dataGen.trackLang(description)
        dataGen.trackLang(details)

        if (!noResearch) {
            research {
                target(loc) {
                    input(getComponent("research_equipment").item(voltage!!))
                    voltage(voltage!!)
                }
            }
        }
    }

    companion object {
        const val RANK_PER_VOLTAGE = 1000

        fun factory(handler: IDataHandler<TechProvider>, parent: Any, loc: ResourceLocation):
            TechBuilder<ResourceLocation> {
            val builder = TechBuilder(loc)
            return builder.onBuild { builder.onRegister(handler, loc) }
        }
    }
}
